styx-client/src/main/java/com/spotify/styx/client/StyxApolloClient.java

```diff
@@ -375,25 +389,29 @@
     });
   }
 
-  private Request decorateRequest(final Request request) {
-    return withOptionalAuth(request.withHeader("User-Agent", STYX_CLIENT_VERSION).withTtl(TTL));
+  private Request decorateRequest(
+      final Request request, final Optional<GoogleCredential> credential) {
+    return withOptionalAuth(request.withHeader("User-Agent", STYX_CLIENT_VERSION).withTtl(TTL),
+        credential);
   }
 
-  private Request withOptionalAuth(final Request request) {
+  private Request withOptionalAuth(
+      final Request request, final Optional<GoogleCredential> credential) {
+    if (!credential.isPresent()) {
+      return request;
+    }
     try {
-      String authToken = new GoogleIdTokenAuth().getToken(this.apiHost.toString());
+      final String authToken = auth.getToken(this.apiHost.toString(), credential.get());
       return request.withHeader("Authorization", "Bearer " + authToken);
-    } catch (IOException e) {
-      // Credential probably not configured. Proceed to invoke API without authentication.
-      return request;
-    } catch (GeneralSecurityException e) {
-      // Credential probably configured wrongly.
+    } catch (IOException | GeneralSecurityException e) {
+      // Credential probably invalid, configured wrongly or the token request failed.
       throw new RuntimeException(e);
     }
   }
 
   private CompletionStage<Response<ByteString>> executeRequest(final Request request) {
-    return client.send(decorateRequest(request)).handle((response, e) -> {
+    final Try<GoogleCredential> credential = Try.of(credentialSupplier);
```

Comment:
Feels a bit weird leaking GoogleCredential in this class. Is it possible to have getToken indicating credential not configured? Eg. throwing a more specific exception than IOException.

Reply:
And when that happens we don’t decorate the request.